lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+        self._db = db
+        self._h_api = h_api
+        self._mailchimp_service = mailchimp_service
+        self._sender = sender
+
     def send_instructor_email_digests(
         self, audience, updated_after, updated_before, override_to_email=None
     ):
         """Send instructor email digests for the given users and timeframe."""
+        annotations = self._h_api.get_annotations(
+            audience, updated_after, updated_before
+        )
+
+        # HAPI.get_annotations() returns an iterable.
+        # Turn it into a tuple because we need to iterate over it multiple times.
+        annotations = tuple(annotations)
+
+        context = DigestContext(self._db, audience, annotations)
+
+        for h_userid in audience:
+            digest = context.instructor_digest(h_userid)
+
+            if not digest["total_annotations"]:
+                # This user has no activity.
+                continue
+
+            unified_user = context.unified_users[h_userid]
+
+            if override_to_email is None:
+                to_email = unified_user.email
+            else:
+                to_email = override_to_email
+
+            if not to_email:
+                # We don't have an email address for this user.
+                continue
+
+            self._mailchimp_service.send_template(
+                "instructor-email-digest",
+                self._sender,
+                recipient=EmailRecipient(to_email, unified_user.display_name),
+                template_vars=digest,
+            )
+
+
+@dataclass
+class UnifiedUser:
+    """All User's for a given h_userid, unified across all ApplicationInstance's."""
+
+    h_userid: str
+    users: Tuple[User]
+    email: Optional[str]
+    display_name: Optional[str]
+
+
+@dataclass
+class UnifiedCourse:
+    """All Course's for a given authority_provided_id, unified across all ApplicationInstance's."""
+
+    authority_provided_id: str
+    title: Optional[str]
+    instructors: Tuple[UnifiedUser]
+    learner_annotations: Tuple[dict]
+
+
+class DigestContext:
+    """A context/helper object for DigestService."""
+
+    def __init__(self, db, audience, annotations):
+        self._db = db
+        self._audience = audience
+        self._annotations = annotations
+        self._unified_users = None
+        self._unified_courses = None
+
+    def instructor_digest(self, h_userid):
+        """
+        Return a digest (dict of template variables) for the given instructor.
+
+        The digest will only include courses in which both:
+
+        1. The user is an instructor
+        2. There are annotations by learners
+        """
+        course_digests = []
+
+        # Remove duplicates from self.unified_courses.values().
+        unified_courses = []
+        for unified_course in self.unified_courses.values():
+            if unified_course not in unified_courses:
+                unified_courses.append(unified_course)
```

Comment:
You can't use a set here because `UnifiedCourse` and `UnifiedUser` aren't hashable.